some ST boards fail to build `samples/sensor/sensor_shell`

**Describe the bug**
`nucleo_h753zi` fails to build `samples/sensor/sensor_shell`, because the `vbat` driver depends on `adc3`, but `adc3` is not enabled in the devicetree.

**To Reproduce**
Steps to reproduce the behavior:
1. `west build -b nucleo_h753zi -p auto zephyr/samples/sensor/sensor_shell`

**Expected behavior**
Sample should compile

**Impact**
This is failing in #72833

**Logs and console output**
```
In file included from /home/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:34,
                 from /home/ycsin/zephyrproject/zephyr/include/zephyr/irq_multilevel.h:15,
                 from /home/ycsin/zephyrproject/zephyr/include/zephyr/devicetree.h:20,
                 from /home/ycsin/zephyrproject/zephyr/include/zephyr/device.h:12,
                 from /home/ycsin/zephyrproject/zephyr/drivers/sensor/st/stm32_vbat/stm32_vbat.c:7:
/home/ycsin/zephyrproject/zephyr/include/zephyr/device.h:91:41: error: '__device_dts_ord_10' undeclared here (not in a function); did you mean '__device_dts_ord_50'?
   91 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                         ^~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:72:26: note: in definition of macro '__DEBRACKET'
   72 | #define __DEBRACKET(...) __VA_ARGS__
      |                          ^~~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:64:9: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   64 |         __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |         ^~~~~~~~~~~~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:59:9: note: in expansion of macro '__COND_CODE'
   59 |         __COND_CODE(_XXXX##_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:180:9: note: in expansion of macro 'Z_COND_CODE_1'
  180 |         Z_COND_CODE_1(_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/devicetree.h:4305:9: note: in expansion of macro 'COND_CODE_1'
 4305 |         COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT),   \
      |         ^~~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/toolchain/common.h:138:23: note: in expansion of macro '_DO_CONCAT'
  138 | #define _CONCAT(x, y) _DO_CONCAT(x, y)
      |                       ^~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/device.h:91:33: note: in expansion of macro '_CONCAT'
   91 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                 ^~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/device.h:228:37: note: in expansion of macro 'DEVICE_NAME_GET'
  228 | #define DEVICE_DT_NAME_GET(node_id) DEVICE_NAME_GET(Z_DEVICE_DT_DEV_ID(node_id))
      |                                     ^~~~~~~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/include/zephyr/device.h:245:34: note: in expansion of macro 'DEVICE_DT_NAME_GET'
  245 | #define DEVICE_DT_GET(node_id) (&DEVICE_DT_NAME_GET(node_id))
      |                                  ^~~~~~~~~~~~~~~~~~
/home/ycsin/zephyrproject/zephyr/drivers/sensor/st/stm32_vbat/stm32_vbat.c:121:24: note: in expansion of macro 'DEVICE_DT_GET'
  121 |                 .adc = DEVICE_DT_GET(DT_INST_IO_CHANNELS_CTLR(inst)),                           \
      |                        ^~~~~~~~~~~~~
/home/ycsin/zephyrproject/build/zephyr/include/generated/devicetree_generated.h:24582:48: note: in expansion of macro 'STM32_VBAT_DEFINE'
24582 | #define DT_FOREACH_OKAY_INST_st_stm32_vbat(fn) fn(0)
      |                                                ^~
/home/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:105:36: note: in expansion of macro 'DT_FOREACH_OKAY_INST_st_stm32_vbat'
  105 | #define UTIL_PRIMITIVE_CAT(a, ...) a##__VA_ARGS__
      |                                    ^
/home/ycsin/zephyrproject/zephyr/drivers/sensor/st/stm32_vbat/stm32_vbat.c:141:1: note: in expansion of macro 'DT_INST_FOREACH_STATUS_OKAY'
  141 | DT_INST_FOREACH_STATUS_OKAY(STM32_VBAT_DEFINE)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~
[74/166] Building C object zephyr/arch/arch/arm/core/cortex_m/CMakeFiles/arch__arm__core__cortex_m.dir/prep_c.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/ycsin/zephyrproject/build
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: 0.16.5-1
 - `main` branch

--- a/drivers/sensor/st/stm32_vbat/Kconfig
+++ b/drivers/sensor/st/stm32_vbat/Kconfig
@@ -7,6 +7,7 @@ config STM32_VBAT
 	bool "STM32 Vbat Sensor"
 	default y
 	depends on DT_HAS_ST_STM32_VBAT_ENABLED
+	depends on DT_HAS_ST_STM32_ADC_ENABLED
 	depends on SOC_FAMILY_STM32 && !SOC_SERIES_STM32F1X
 	select ADC
 	help

--- a/drivers/sensor/st/stm32_vbat/stm32_vbat.c
+++ b/drivers/sensor/st/stm32_vbat/stm32_vbat.c
@@ -101,6 +101,11 @@ static int stm32_vbat_init(const struct device *dev)
 
 	k_mutex_init(&data->mutex);
 
+	if (data->adc == NULL) {
+		LOG_ERR("ADC is not enabled");
+		return -ENODEV;
+	}
+
 	if (!device_is_ready(data->adc)) {
 		LOG_ERR("Device %s is not ready", data->adc->name);
 		return -ENODEV;
@@ -116,9 +121,13 @@ static int stm32_vbat_init(const struct device *dev)
 	return 0;
 }
 
+#define STM32_VBAT_GET_ADC_OR_NULL(inst)                                                           \
+	COND_CODE_1(DT_NODE_HAS_STATUS(DT_INST_IO_CHANNELS_CTLR(inst), okay),                      \
+		    (DEVICE_DT_GET(DT_INST_IO_CHANNELS_CTLR(inst))), (NULL))
+
 #define STM32_VBAT_DEFINE(inst)									\
 	static struct stm32_vbat_data stm32_vbat_dev_data_##inst = {				\
-		.adc = DEVICE_DT_GET(DT_INST_IO_CHANNELS_CTLR(inst)),				\
+		.adc = STM32_VBAT_GET_ADC_OR_NULL(inst),					\
 		.adc_base = (ADC_TypeDef *)DT_REG_ADDR(DT_INST_IO_CHANNELS_CTLR(0)),		\
 		.adc_cfg = {									\
 			.gain = ADC_GAIN_1,							\
@@ -136,6 +145,6 @@ static int stm32_vbat_init(const struct device *dev)
 	SENSOR_DEVICE_DT_INST_DEFINE(inst, stm32_vbat_init, NULL,				\
 			      &stm32_vbat_dev_data_##inst, &stm32_vbat_dev_config_##inst,	\
 			      POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY,				\
-			      &stm32_vbat_driver_api);						\
+			      &stm32_vbat_driver_api);
 
 DT_INST_FOREACH_STATUS_OKAY(STM32_VBAT_DEFINE)

